app/services/DriveService.scala

```diff
@@ -34,34 +35,90 @@ object DriveService {
       .setTeamDriveId(Config.teamDriveId)
 
     request.execute().getFiles.asScala.toList
-      .map(file => { // extract this into a pure function we can test
-
-        // raw model file
-        // function transforms
-        // output target models
-
-        // The inline Try will handle nulls returned from Java library
-        val customProps = Try {
-          file.getProperties.asScala.toMap
-        }.toOption.getOrElse(Map.empty)
-        val thumbnailLink = Try {
-          file.getThumbnailLink
-        }.toOption.getOrElse("")
-        val exportLinks = Try {
-          file.getExportLinks.asScala.toMap
-        }.toOption.getOrElse(Map.empty)
-
-        File(
-          id = file.getId,
-          title = file.getName,
-          output = file.getMimeType,
-          outputPreview = file.getWebViewLink, // TODO - returning editable links - need to process to preview
-          outputDownload = file.getWebContentLink, // TODO - returning nulls
-          outputThumbnail = thumbnailLink, // TODO - returning 404s
-          exportLinks = exportLinks,
-          customProperties = customProps
-        )
-      })
+      .map(file =>
+        // transformToResearchRecord(extractRawFile(file)) // define functions with def
+        (extractRawFile andThen transformToResearchRecord)(file) // define functions with val
+    )
+  }
+
+  val extractRawFile = (fileStream: File) => {
+    val customProps = Try {
+      fileStream.getProperties.asScala.toMap
+    }.toOption.getOrElse(Map.empty)
+    val exportLinks = Try {
+      fileStream.getExportLinks.asScala.toMap
+    }.toOption.getOrElse(Map.empty)
+
+    DriveFile(
+      id = fileStream.getId,
+      title = fileStream.getName,
+      output = fileStream.getMimeType,
+      outputPreview = fileStream.getWebViewLink,
+      outputDownload = fileStream.getWebContentLink, // TODO - returning nulls
+      exportLinks = exportLinks,
+      customProperties = customProps
```

Comment:
this is fine, but I notice that most of the fields in this class are String.  This is OK in this relatively simple logic, but bear in mind the scala compiler won't be able to tell you if you mix them up or copy and paste error.  A bit like your question about Quarter type earlier, you get help from the compiler checking even for more native values than Quarter.

In effect, in code with more tricky logic it's worth having a data transfer model as you have already, plus a domain model which uses value classes and other specific data types.
e.g.
```scala
case class DriveTitle(value: String) extends AnyVal
```
This does mean you have to have a mapping layer between your low level transfer model and your domain model, so it isn't something worthwhile in very simple logic, but it does help if you want to write and test logic without having to depend directly on the external model.

Happy to explain more.